library/std/src/sys/hermit/thread_local_dtor.rs

```diff
@@ -5,23 +5,22 @@
 // The this solution works like the implementation of macOS and
 // doesn't additional OS support
 
-use crate::mem;
+use crate::cell::RefCell;
 
 #[thread_local]
-static mut DTORS: Vec<(*mut u8, unsafe extern "C" fn(*mut u8))> = Vec::new();
+static DTORS: RefCell<Vec<(*mut u8, unsafe extern "C" fn(*mut u8))>> = RefCell::new(Vec::new());
 
 pub unsafe fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
-    let list = &mut DTORS;
-    list.push((t, dtor));
+    DTORS.try_borrow_mut().expect("global allocator may not use TLS").push((t, dtor));
```

Comment:
You are correct, panicking will allocate. I'm not sure whether it's worth it to abort here, because those allocations could also succeed; so we would unnecessarily limit the allowed behaviour (if the allocator catches the panic, for instance). In the case of this PR, I don't see how the behaviour could be allowed, hence the panic.